pkg/controller/core/resource_controller.go

```diff
@@ -54,10 +54,60 @@ const (
+// of the reconcile function explicit.
 var (
-	Result        = reconcile.Result{}
-	ResultRequeue = reconcile.Result{Requeue: true}
+	// DeprecatedResultDone indicates that no further reconcilations need to be
+	// queued for the object under reconciliation. It is deprecated and should
+	// be replaced with a more explicitly named reconcile result.
+	DeprecatedResultDone = reconcile.Result{Requeue: false}
+
+	// DeprecatedResultRequeue indicates that further reconciliations need to be
+	// queued for the object under reconciliattion. It is deprecated and should
+	// be replaced with a more explicitly named reconcile result.
+	DeprecatedResultRequeue = reconcile.Result{Requeue: true}
+
+	// ResultDirty should be returned when the reconciler has altered the object
+	// under reconciliation. When returning this result the reconciler must also
+	// submit the updated object to the API server, thus triggering an implicit
+	// reconciliation of the changed object.
+	ResultDirty = reconcile.Result{Requeue: false}
+
+	// ResultDeleted should be returned when the object under reconciliation has
+	// been deleted, i.e. after the reconciler has removed any finalizers.
+	ResultDeleted = ResultDirty
+
+	// ResultClean should be returned when the reconciler has not altered the
+	// object under reconciliation, and is not waiting for a pending operation
+	// to complete.
+	ResultClean = reconcile.Result{RequeueAfter: RequeueOnSuccess}
+
+	// ResultWaiting should be returned when the reconciler has not altered the
+	// object under reconciliation, but is waiting for a pending operation to
+	// complete, for example waiting for the managed cloud resource the object
+	// represents to transition from 'Creating' to 'Available'.
+	ResultWaiting = reconcile.Result{RequeueAfter: RequeueOnWait}
+
+	// ResultNotFound should be returned when the object under reconciliation no
+	// longer exists when the reconcile function is invoked.
+	ResultNotFound = reconcile.Result{Requeue: false}
+
+	// ResultError should be returned if the reconciler knows it will return an
+	// error, thus triggering an implicit reconciliation.
+	ResultError = reconcile.Result{Requeue: false}
+
+	// ResultFailed should be returned if the reconciler had reason to set the
+	// object under reconciliation's status to 'failed'. This typically means
+	// an error was encountered and reflected in the object's status instead of
+	// being returned by the reconciler. In this scenario we want to requeue
+	// with exponential backoff. The eagle eyed reader may note that returning
+	// ResultFailed will duplicate the implicit reconcile triggered by setting
+	// the object's status to failed. We return ResultFailed rather than
+	// ResultDirty to account for the case in which the object was already, and
+	// remains, in state failed and thus does not trigger an implicit
+	// reconciliation.
+	ResultFailed = reconcile.Result{Requeue: true}
```

Comment:
I am not :100: about the necessity of these constants. Different controller/types may need to have different or more granular values and use cases. On another hand: `reconcile.Result{}` - is self-explanatory, where as `ResultDelete`->`ResultDirty`->`reconcile.Result{}` arguably not so much. I am not saying this is a show stopper for me. At the same time seems a bit excessive.

Reply:
Personally speaking I'm still learning things and being surprised by the semantics of controller-runtime reconcile results as recently as today, after 2-3 months of working on Crossplane, so I find it difficult to understand how `reconcile.Result{}` is self-explanatory.

Yes, if you have a half decent IDE you can click through to the symbol definition somewhere in another file and see that there is a `Requeue` boolean field. If you're familiar with Go you'll know the zero value of that omitted field is false. If you've read our forthcoming design document you'll know that it makes sense in certain circumstances to return `reconcile.Result{}` where it otherwise seems unintuitive because an error or `client.Update()` call triggers an implicit reconcile.

On the other hand you could see `ResultNotFound` and get a pretty good idea that it's a result returned when an object is not found. I'll grant it isn't perfect, but I think the strength of using constants in this case is more in their docstrings than the constants themselves. If I were reading this code I could quickly click through to `ResultNotFound` (or lookup its godoc) and hopefully learn enough about working with it that I'd get things mostly right as a newcomer to this project.

I don't feel strongly about this exact set of constants, or the exact names of these constants, and I do not expect these constants to be ideal for 100% of cases, but I do think these or something like them could improve the readability of our controllers in ~80% of cases.

Reply:
Given this is a little controversial and affects more than just complex workloads I'm leaning toward removing the commit from this specific PR so we can figure out what we want to do in a separate PR.

Reply:
I think you've made a good case for these constants usage, so I am fine with this change here.

Reply:
I'll break it out into another PR, mostly because while I do think we should have some kind of constants I am not 💯 on whether this naming scheme is the best one.

Reply:
Moved to https://github.com/crossplaneio/crossplane/pull/424